File: web/backend/app/models/simulation.py
# ABOUTME: Pydantic models for simulation parameters and API responses
# ABOUTME: Defines request/response schemas for the 3D Physarum simulation API

from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from enum import Enum


class SmoothingType(str, Enum):
    """Smoothing algorithm types for smooth surface generation."""
    laplacian = "laplacian"
    taubin = "taubin"
    feature_preserving = "feature_preserving"
    boundary_outline = "boundary_outline"


class SimulationStatus(str, Enum):
    """Status of a simulation job."""
    pending = "pending"
    running = "running"
    completed = "completed"
    failed = "failed"
    cancelled = "cancelled"


class SimulationParameters(BaseModel):
    """Parameters for starting a new simulation."""
    
    # Simulation parameters
    width: int = Field(default=100, ge=1, le=2000, description="Grid width in pixels")
    height: int = Field(default=100, ge=1, le=2000, description="Grid height in pixels")
    actors: int = Field(default=50, ge=1, le=10000, description="Number of Physarum actors")
    steps: int = Field(default=100, ge=1, le=10000, description="Number of simulation steps")
    decay: float = Field(default=0.01, ge=0.0, le=1.0, description="Trail decay rate")
    view_radius: int = Field(default=3, ge=0, le=20, description="Actor sensing radius")
    view_distance: int = Field(default=10, ge=0, le=50, description="Actor sensing distance")
    speed: float = Field(default=1.0, ge=0.0, le=10.0, description="Actor movement speed")
    speed_min: Optional[float] = Field(default=None, ge=0.0, le=10.0, description="Minimum speed for initial actors")
    speed_max: Optional[float] = Field(default=None, ge=0.0, le=10.0, description="Maximum speed for initial actors")
    spawn_speed_randomization: float = Field(default=0.2, ge=0.0, le=1.0, description="Factor for randomizing spawned actor speeds")
    initial_diameter: float = Field(default=20.0, ge=1.0, le=1000.0, description="Initial circle diameter for actor placement")
    death_probability: float = Field(default=0.001, ge=0.0, le=1.0, description="Age-based death probability per step")
    spawn_probability: float = Field(default=0.005, ge=0.0, le=1.0, description="Probability of spawning new actors")
    diffusion_rate: float = Field(default=0.0, ge=0.0, le=1.0, description="Pheromone diffusion rate")
    direction_deviation: float = Field(default=1.57, ge=0.0, le=3.14159, description="Maximum direction deviation in radians")
    image: Optional[str] = Field(default=None, description="Base64 encoded JPEG image for initial actor placement")
    
    # 3D model parameters
    smooth: bool = Field(default=False, description="Use smooth surface generation (marching cubes)")
    layer_height: float = Field(default=1.0, ge=0.1, le=10.0, description="Height of each layer in 3D model")
    threshold: float = Field(default=0.1, ge=0.0, le=1.0, description="Minimum trail strength for 3D inclusion")
    layer_frequency: int = Field(default=5, ge=1, le=100, description="Capture layer every N steps")
    smoothing_iterations: int = Field(default=2, ge=0, le=20, description="Number of smoothing iterations")
    smoothing_type: SmoothingType = Field(default=SmoothingType.taubin, description="Type of smoothing algorithm")
    taubin_lambda: float = Field(default=0.5, gt=0.0, lt=1.0, description="Taubin smoothing lambda parameter")
    taubin_mu: float = Field(default=-0.52, gt=-1.0, lt=0.0, description="Taubin smoothing mu parameter")
    preserve_features: bool = Field(default=False, description="Preserve sharp features during smoothing")
    feature_angle: float = Field(default=60.0, gt=0.0, lt=180.0, description="Feature edge angle threshold in degrees")
    mesh_quality: bool = Field(default=False, description="Show detailed mesh quality metrics")
    background: bool = Field(default=False, description="Add a solid rectangular background")
    background_depth: float = Field(default=2.0, gt=0.0, le=50.0, description="Depth/thickness of the background layer")
    background_margin: float = Field(default=0.05, ge=0.0, le=1.0, description="Background margin as fraction of bounds")
    background_border: bool = Field(default=False, description="Add a raised border around the background edges")
    border_height: float = Field(default=1.0, gt=0.0, le=50.0, description="Height of the border walls")
    border_thickness: float = Field(default=0.5, gt=0.0, le=10.0, description="Thickness of the border walls")
    
    # Output parameters
    output: str = Field(default="physarum_3d_model.stl", description="Output STL filename")
    
    @validator('speed_min', 'speed_max')
    def validate_speeds(cls, v, values):
        """Validate speed min/max relationships."""
        if v is not None:
            if v <= 0:
                raise ValueError("Speed values must be positive")
            if 'speed_min' in values and 'speed_max' in values:
                if values.get('speed_min') is not None and v is not None:
                    if values['speed_min'] > v:
                        raise ValueError("Speed minimum must be less than or equal to speed maximum")
        return v


class SimulationRequest(BaseModel):
    """Request to start a new simulation."""
    parameters: SimulationParameters
    
    
class SimulationResponse(BaseModel):
    """Response when starting a new simulation."""
    job_id: str
    status: SimulationStatus
    message: str


class ProgressUpdate(BaseModel):
    """Real-time progress update for a simulation."""
    job_id: str
    step: int
    total_steps: int
    layers_captured: int
    actor_count: int
    max_trail: float
    mean_trail: float
    estimated_completion_time: Optional[float] = None
    timestamp: float


class SimulationStatusResponse(BaseModel):
    """Response for simulation status queries."""
    job_id: str
    status: SimulationStatus
    progress: Optional[ProgressUpdate] = None
    error_message: Optional[str] = None
    started_at: Optional[float] = None
    completed_at: Optional[float] = None


class MeshQualityMetrics(BaseModel):
    """Mesh quality metrics for completed simulations."""
    vertex_count: int
    face_count: int
    volume: float
    surface_area: float
    is_watertight: bool
    is_winding_consistent: bool
    print_ready: bool
    issues: List[str]


class SimulationResult(BaseModel):
    """Result of a completed simulation."""
    job_id: str
    status: SimulationStatus
    parameters: SimulationParameters
    files: Dict[str, str]  # file_type -> relative_path
    statistics: Dict[str, Any]
    mesh_quality: Optional[MeshQualityMetrics] = None
    completed_at: float
    file_sizes: Dict[str, int]  # file_type -> size_in_bytes


class ErrorResponse(BaseModel):
    """Error response for API failures."""
    error: str
    message: str
    job_id: Optional[str] = None
    details: Optional[Dict[str, Any]] = None

File: web/frontend/src/types/simulation.ts
// ABOUTME: TypeScript interfaces and types for simulation parameters and API responses
// ABOUTME: Defines the structure of all data passed between frontend and backend

export interface SimulationParameters {
  // Simulation Parameters
  width: number;
  height: number;
  actors: number;
  steps: number;
  decay: number;
  viewRadius: number;
  viewDistance: number;
  speed: number;
  speedMin?: number;
  speedMax?: number;
  spawnSpeedRandomization: number;
  initialDiameter: number;
  deathProbability: number;
  spawnProbability: number;
  diffusionRate: number;
  directionDeviation: number;
  image?: string; // Image path for initial actor placement

  // 3D Model Parameters
  smooth: boolean;
  layerHeight: number;
  threshold: number;
  layerFrequency: number;
  smoothingIterations: number;
  smoothingType: 'laplacian' | 'taubin' | 'feature_preserving' | 'boundary_outline';
  taubinLambda: number;
  taubinMu: number;
  preserveFeatures: boolean;
  featureAngle: number;
  meshQuality: boolean;
  background: boolean;
  backgroundDepth: number;
  backgroundMargin: number;
  backgroundBorder: boolean;
  borderHeight: number;
  borderThickness: number;

  // Output Parameters
  output: string;
  quiet: boolean;
  verbose: boolean;
}

export interface SimulationStatus {
  jobId: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';
  currentStep: number;
  totalSteps: number;
  capturedLayers: number;
  actorCount: number;
  trailStats: {
    maxTrail: number;
    meanTrail: number;
  };
  estimatedTimeRemaining?: number;
  progress: number; // 0-100
  error?: string;
}

export interface SimulationResult {
  jobId: string;
  stlPath: string;
  jsonPath: string;
  previewPath: string;
  fileSize: number;
  parameters: SimulationParameters;
  meshMetrics?: {
    vertexCount: number;
    faceCount: number;
    volume: number;
    surfaceArea: number;
    isWatertight: boolean;
    isWindingConsistent: boolean;
    printReady: boolean;
    issues: string[];
  };
  completedAt: string;
}

export interface ParameterPreset {
  name: string;
  description: string;
  parameters: Partial<SimulationParameters>;
  estimatedTime?: string;
  complexity?: 'low' | 'medium' | 'high' | 'very_high';
  tags?: string[];
}

export const DEFAULT_PARAMETERS: SimulationParameters = {
  // Simulation Parameters
  width: 100,
  height: 100,
  actors: 50,
  steps: 100,
  decay: 0.01,
  viewRadius: 3,
  viewDistance: 10,
  speed: 1.0,
  spawnSpeedRandomization: 0.2,
  initialDiameter: 20.0,
  deathProbability: 0.001,
  spawnProbability: 0.005,
  diffusionRate: 0.0,
  directionDeviation: 1.57,

  // 3D Model Parameters
  smooth: false,
  layerHeight: 1.0,
  threshold: 0.1,
  layerFrequency: 5,
  smoothingIterations: 2,
  smoothingType: 'taubin',
  taubinLambda: 0.5,
  taubinMu: -0.52,
  preserveFeatures: false,
  featureAngle: 60.0,
  meshQuality: false,
  background: false,
  backgroundDepth: 2.0,
  backgroundMargin: 0.05,
  backgroundBorder: false,
  borderHeight: 1.0,
  borderThickness: 0.5,

  // Output Parameters
  output: 'physarum_3d_model.stl',
  quiet: false,
  verbose: false,
};

export interface PresetCategory {
  name: string;
  description: string;
  presets: ParameterPreset[];
}

export const PARAMETER_PRESETS: ParameterPreset[] = [
  {
    name: 'Fast Preview',
    description: 'Quick generation for testing parameters (~1-2 min)',
    parameters: {
      width: 50,
      height: 50,
      actors: 25,
      steps: 50,
      layerFrequency: 10,
      smooth: false,
      output: 'fast_preview.stl',
    },
  },
  {
    name: 'Balanced Quality',
    description: 'Good quality with reasonable time (~5-10 min)',
    parameters: {
      width: 100,
      height: 100,
      actors: 75,
      steps: 150,
      smooth: true,
      smoothingIterations: 2,
      layerFrequency: 5,
      meshQuality: false,
      output: 'balanced_quality.stl',
    },
  },
  {
    name: 'High Quality',
    description: 'Detailed model with smooth surfaces (~15-30 min)',
    parameters: {
      width: 200,
      height: 200,
      actors: 100,
      steps: 300,
      smooth: true,
      smoothingIterations: 3,
      layerFrequency: 3,
      meshQuality: true,
      preserveFeatures: true,
      featureAngle: 45,
      output: 'high_quality.stl',
    },
  },
  {
    name: 'Complex Structure',
    description: 'Dense network with intricate patterns (~20-45 min)',
    parameters: {
      width: 150,
      height: 150,
      actors: 200,
      steps: 400,
      decay: 0.005,
      spawnProbability: 0.01,
      diffusionRate: 0.1,
      smooth: true,
      smoothingIterations: 2,
      layerFrequency: 4,
      output: 'complex_structure.stl',
    },
  },
  {
    name: 'Organic Growth',
    description: 'Natural branching patterns (~10-20 min)',
    parameters: {
      width: 120,
      height: 120,
      actors: 30,
      steps: 250,
      decay: 0.02,
      viewRadius: 5,
      viewDistance: 15,
      speed: 1.5,
      spawnProbability: 0.003,
      directionDeviation: 0.8,
      smooth: true,
      smoothingIterations: 2,
      output: 'organic_growth.stl',
    },
  },
  {
    name: 'Printable Model',
    description: 'Optimized for 3D printing with background (~10-15 min)',
    parameters: {
      width: 100,
      height: 100,
      actors: 60,
      steps: 180,
      smooth: true,
      smoothingIterations: 3,
      layerFrequency: 5,
      meshQuality: true,
      background: true,
      backgroundDepth: 2.0,
      backgroundMargin: 0.1,
      backgroundBorder: true,
      borderHeight: 1.5,
      borderThickness: 1.0,
      threshold: 0.15,
      output: 'printable_model.stl',
    },
  },
  {
    name: 'Miniature Detail',
    description: 'Fine details for small models (~5-10 min)',
    parameters: {
      width: 80,
      height: 80,
      actors: 40,
      steps: 120,
      decay: 0.015,
      viewRadius: 2,
      viewDistance: 8,
      speed: 0.8,
      layerHeight: 0.5,
      layerFrequency: 3,
      smooth: true,
      smoothingIterations: 4,
      threshold: 0.2,
      output: 'miniature_detail.stl',
    },
  },
  {
    name: 'Artistic Abstract',
    description: 'Creative patterns with high randomness (~15-25 min)',
    parameters: {
      width: 140,
      height: 140,
      actors: 80,
      steps: 320,
      decay: 0.008,
      spawnProbability: 0.008,
      spawnSpeedRandomization: 0.5,
      directionDeviation: 1.2,
      diffusionRate: 0.05,
      smooth: true,
      smoothingIterations: 2,
      layerFrequency: 4,
      output: 'artistic_abstract.stl',
    },
  },
];

export const PRESET_CATEGORIES: PresetCategory[] = [
  {
    name: 'Quick Start',
    description: 'Fast presets for testing and preview',
    presets: PARAMETER_PRESETS.filter(p => 
      ['Fast Preview', 'Balanced Quality'].includes(p.name)
    ),
  },
  {
    name: 'Quality Focus',
    description: 'High-quality models for final output',
    presets: PARAMETER_PRESETS.filter(p => 
      ['High Quality', 'Printable Model'].includes(p.name)
    ),
  },
  {
    name: 'Creative Patterns',
    description: 'Unique and artistic configurations',
    presets: PARAMETER_PRESETS.filter(p => 
      ['Complex Structure', 'Organic Growth', 'Artistic Abstract'].includes(p.name)
    ),
  },
  {
    name: 'Specialized',
    description: 'Purpose-built configurations',
    presets: PARAMETER_PRESETS.filter(p => 
      ['Miniature Detail'].includes(p.name)
    ),
  },
];